apollo-router/src/plugins/rhai.rs

```diff
@@ -1,105 +1,240 @@
-                        .extensions(Object::default())
-                        .build();
-                }
-                if function_found {
-                    let rhai_context = match this.run_rhai_script(
-                        $function_name,
-                        response.context,
-                        response.response.headers().clone(),
-                    ) {
-                        Ok(res) => res,
-                        Err(err) => {
-                            let context = Context::new();
-                            context
-                                .insert(CONTEXT_ERROR, err)
-                                .expect("error is always a string; qed");
-
-                            return $response_ty::builder()
-                                .context(context)
-                                .extensions(Object::default())
-                                .build();
-                        }
-                    };
-                    response.context = rhai_context.context;
-                    *response.response.headers_mut() = rhai_context.headers;
+use std::str::FromStr;
+use std::{
+    ops::ControlFlow,
+    path::PathBuf,
+    sync::{Arc, Mutex},
+};
+use tower::{util::BoxService, BoxError, ServiceBuilder, ServiceExt};
 
-                    response.response.headers_mut().remove(CONTENT_LENGTH);
-                }
+pub trait Accessor<Access>: Send {
+    fn accessor(&self) -> &Access;
 
-                response
-            })
-            .boxed();
-    };
+    fn accessor_mut(&mut self) -> &mut Access;
 }
 
-macro_rules! handle_error {
-    ($call: expr, $req: expr) => {
-        match $call {
-            Ok(res) => res,
-            Err(err) => {
-                $req.context
-                    .insert(CONTEXT_ERROR, err)
-                    .expect("we manually created error string; qed");
-                return $req;
-            }
-        }
-    };
+trait OptionDance<T> {
+    fn with_mut<R>(&self, f: impl FnOnce(&mut T) -> R) -> R;
+
+    fn replace(&self, f: impl FnOnce(T) -> T);
+
+    fn take_unwrap(self) -> T;
 }
 
-use rhai::plugin::*; // a "prelude" import for macros
+impl<T> OptionDance<T> for Arc<Mutex<Option<T>>> {
+    fn with_mut<R>(&self, f: impl FnOnce(&mut T) -> R) -> R {
+        let mut guard = self.lock().expect("poisoned mutex");
+        f(guard.as_mut().expect("re-entrant option dance"))
+    }
+
+    fn replace(&self, f: impl FnOnce(T) -> T) {
+        let mut guard = self.lock().expect("poisoned mutex");
+        *guard = Some(f(guard.take().expect("re-entrant option dance")))
+    }
+
+    fn take_unwrap(self) -> T {
+        match Arc::try_unwrap(self) {
+            Ok(mutex) => mutex.into_inner().expect("poisoned mutex"),
+
+            // TODO: Should we assume the Arc refcount is 1
+            // and use `try_unwrap().expect("shared ownership")` instead of this fallback ?
+            Err(arc) => arc.lock().expect("poisoned mutex").take(),
```

Comment:
I think the expectation around a `take_unwrap` function would be that it does `try_unwrap().expect("shared ownership")`

Reply:
Does Rhai have global variables? Would it be possible for a Rhai function to stash somewhere a strong reference that lasts longer that the call, to a value passed to it as an argument?

Reply:
Yes rhai has a concept of global variables thanks to the concept of "context" if I recall correctly

Reply:
So it’s possible that a script would keep a reference to a `service`, so that the `Arc` refcount would be greater than 1 so `try_unwrap` would fail. How do we want to handle that situation?

Reply:
We actually control the "Scope" (that's the name in rhai, not context) and we can prevent this from happening by strictly controlling what can persist in scope. At the moment, we don't do that, but I think it would be fairly straightforward to add.

Reply:
@SimonSapin I just had a (perhaps) better idea. We could require that the various _service functions return the service (possibly decorated or modified) that we pass in. Rhai doesn't have a way to take a reference for a variable, so this would eliminate the possibility that an additional reference exists to the service.